src/common.cpp

```diff
@@ -1,10 +1,93 @@
 #include "common.hh"
+#include "zarr.hh"
 
 #include "platform.h"
 
 #include <cmath>
+#include <thread>
 
 namespace common = acquire::sink::zarr::common;
+common::ThreadPool::ThreadPool(size_t n_threads,
+                               std::function<void(const std::string&)> err)
+  : contexts_{ n_threads }
+{
+    if (n_threads == 0) {
+        throw std::runtime_error("Cannot create thread pool with 0 threads.");
+    }
+
+    for (auto& ctx_ : contexts_) {
+        ctx_.should_stop = false;
+        ctx_.ready = true;
+        ctx_.thread = std::thread([this, ctx = &ctx_] { thread_worker_(ctx); });
+    }
+}
+
+common::ThreadPool::~ThreadPool() noexcept
+{
+    // spin down threads
+    for (auto& ctx : contexts_) {
+        ctx.should_stop = true;
+        ctx.cv.notify_one();
+        ctx.thread.join();
+    }
+}
+
+void
+common::ThreadPool::push_to_job_queue(JobT&& job)
+{
+    std::scoped_lock lock(jobs_mutex_);
+    jobs_.push(std::move(job));
+}
+
+std::optional<common::ThreadPool::JobT>
+common::ThreadPool::pop_from_job_queue_() noexcept
+{
+    std::scoped_lock lock(jobs_mutex_);
+    if (jobs_.empty()) {
+        return std::nullopt;
+    }
+
+    auto job = jobs_.front();
+    jobs_.pop();
+    return job;
+}
+
+void
+common::ThreadPool::thread_worker_(ThreadContext* ctx)
+{
+    using namespace std::chrono_literals;
+
+    TRACE("Worker thread starting.");
+    if (nullptr == ctx) {
+        LOGE("Null context passed to worker thread.");
+        return;
+    }
+
+    while (true) {
+        std::unique_lock lock(ctx->mutex);
+        ctx->cv.wait_for(lock, 1ms, [&] { return ctx->should_stop; });
+
+        if (ctx->should_stop) {
+            break;
+        }
+
+        // FIXME: wait
+        if (auto job = pop_from_job_queue_(); job.has_value()) {
+            ctx->ready = false;
+            if (std::string err_msg; !job.value()(err_msg)) {
+                error_handler_(err_msg);
+            }
```

Comment:
Just pointing out the non-standard if statement here. I can't tell whether I like yours or the suggestion more.

Maybe only invoke the `error_handler_` on a non-empty `err_msg`?

```suggestion
            {
               std::string err_msg;
               if (!job.value()(err_msg)) {
                   error_handler_(err_msg);
               }
            }
```

Reply:
Functionally they're the same.

Why would we only invoke the error handler when `err_msg` is nonempty? The job returns a bool, and if it returns `false` the Zarr writer should fail on the next `append` call. The error message is just diagnostic.